internal/controller/delayed_restart_manager_test.go

```diff
@@ -0,0 +1,808 @@
+	t.Parallel()
+
+	logger := logr.Discard()
+	manager := NewDelayedRestartManager(logger)
+	defer manager.Shutdown()
+
+	ctx := context.Background()
+	target := TargetKey{
+		Kind:      "Deployment",
+		Name:      "test-deployment",
+		Namespace: "default",
+	}
+
+	executed := make(chan struct{})
+	restartFunc := func(ctx context.Context) error {
+		close(executed)
+
+		return nil
+	}
+
+	// Schedule immediate restart (delay=0)
+	finalDelay, isNew, err := manager.ScheduleRestart(ctx, target, 0, restartFunc)
+
+	if err != nil {
+		t.Errorf("ScheduleRestart() with delay=0 returned error: %v", err)
+	}
+
+	if !isNew {
+		t.Error("ScheduleRestart() with delay=0 should return isNew=true")
+	}
+
+	if finalDelay != 0 {
+		t.Errorf("ScheduleRestart() with delay=0 returned finalDelay=%v, want 0", finalDelay)
+	}
+
+	// Wait for execution with timeout
+	select {
+	case <-executed:
+		// Success - restart executed immediately
+	case <-time.After(200 * time.Millisecond):
+		t.Error("ScheduleRestart() with delay=0 did not execute within 200ms")
+	}
+}
+
+func TestDelayedRestartManager_ScheduleRestart_Delayed(t *testing.T) {
+	t.Parallel()
+
+	logger := logr.Discard()
+	manager := NewDelayedRestartManager(logger)
+	defer manager.Shutdown()
+
+	ctx := context.Background()
+	target := TargetKey{
+		Kind:      "Deployment",
+		Name:      "test-deployment",
+		Namespace: "default",
+	}
+
+	startTime := time.Now()
+	executed := make(chan time.Time)
+	restartFunc := func(ctx context.Context) error {
+		executed <- time.Now()
+
+		return nil
+	}
+
+	delay := 100 * time.Millisecond
+
+	// Schedule delayed restart
+	finalDelay, isNew, err := manager.ScheduleRestart(ctx, target, delay, restartFunc)
+
+	if err != nil {
+		t.Errorf("ScheduleRestart() returned error: %v", err)
+	}
+
+	if !isNew {
+		t.Error("ScheduleRestart() should return isNew=true")
+	}
+
+	if finalDelay != delay {
+		t.Errorf("ScheduleRestart() returned finalDelay=%v, want %v", finalDelay, delay)
+	}
+
+	// Check that restart is scheduled
+	if !manager.IsScheduled(target) {
+		t.Error("IsScheduled() returned false after scheduling")
+	}
+
+	// Wait for execution with timeout
+	select {
+	case executedAt := <-executed:
+		elapsed := executedAt.Sub(startTime)
+		if elapsed < delay {
+			t.Errorf("Restart executed too early, elapsed=%v, expected at least %v", elapsed, delay)
+		}
+		if elapsed > delay+50*time.Millisecond {
+			t.Errorf("Restart executed too late, elapsed=%v, expected around %v", elapsed, delay)
+		}
+	case <-time.After(delay + 200*time.Millisecond):
+		t.Error("ScheduleRestart() did not execute within expected time")
+	}
+
+	// Check that task is removed after execution
+	time.Sleep(10 * time.Millisecond)
+	if manager.IsScheduled(target) {
+		t.Error("IsScheduled() returned true after execution")
+	}
+}
+
+//nolint:cyclop // Test function complexity is acceptable for comprehensive testing
+func TestDelayedRestartManager_ScheduleRestart_MultipleRulesHighestDelayWins(t *testing.T) {
```

Comment:
Implementation differs from issue specification. If a restart is already scheduled, the scheduled reatart should not be prolonged because this might defer the restart to far away.